### PR TITLE
chat exec pluginapi

### PR DESCRIPTION
**Pull request by DerZombiiie**

### CODE REVIEW

**Comment by DerZombiiie** [comment]
A Exec function on client conn to send chat messages as if the client would have send it

File: chat.go [hunk 1]
```diff
@@ -13,6 +13,25 @@ import (
 // about a chat command that's taking long to execute.
 var ChatCmdTimeout = 10 * time.Second
 
+// Exec handles a string as if it where send by a client.
```
Comment (HimbeerserverDE) [inline chat.go:16]: Aaaaargh english
And why is it called Exec?
What about DoChatMsg or something like that?

File: chat.go [hunk 2]
```diff
@@ -13,6 +13,25 @@ import (
...
+// bool   is if msg is getting forwarded to the minetest server
+func (cc *ClientConn) Exec(msg string) (string, bool) {
+	pkt := &mt.ToSrvChatMsg{
+		Msg: msg,
+	}
+
+	str, fwd := onChatMsg(cc, pkt)
+
+	if fwd {
+		cc.server().Send(*&mt.Pkt{
```
Comment (HimbeerserverDE) [inline chat.go:27]: SendCmd does the exact same thing
Also dafuq is this *& stuff?!
Btw if SendCmd doesn’t exist, implement the wrapper for mt.Peer.SendCmd yourself

File: chat.go [hunk 3]
```diff
@@ -13,6 +13,25 @@ import (
...
+// Exec handles a string as if it where send by a client.
+// string is imediate returnvalue of chatcommand (if applicable)
+// bool   is if msg is getting forwarded to the minetest server
+func (cc *ClientConn) Exec(msg string) (string, bool) {
+	pkt := &mt.ToSrvChatMsg{
+		Msg: msg,
+	}
+
+	str, fwd := onChatMsg(cc, pkt)
+
```
Comment (HimbeerserverDE) [inline chat.go:25]: Delete this line please

File: chat.go [hunk 4]
```diff
@@ -13,6 +13,25 @@ import (
...
 
+// Exec handles a string as if it where send by a client.
+// string is imediate returnvalue of chatcommand (if applicable)
+// bool   is if msg is getting forwarded to the minetest server
+func (cc *ClientConn) Exec(msg string) (string, bool) {
+	pkt := &mt.ToSrvChatMsg{
+		Msg: msg,
+	}
+
+	str, fwd := onChatMsg(cc, pkt)
```
Comment (HimbeerserverDE) [inline chat.go:24]: This can block, make it concurrent just like regular chat handling

File: chat.go [hunk 5]
```diff
@@ -13,6 +13,25 @@ import (
 // about a chat command that's taking long to execute.
 var ChatCmdTimeout = 10 * time.Second
 
+// Exec handles a string as if it where send by a client.
+// string is imediate returnvalue of chatcommand (if applicable)
+// bool   is if msg is getting forwarded to the minetest server
+func (cc *ClientConn) Exec(msg string) (string, bool) {
+	pkt := &mt.ToSrvChatMsg{
```
Comment (HimbeerserverDE) [inline chat.go:20]: You can actually just write &mt.ToSrvChatMsg{msg} here

File: chat.go [hunk 6]
```diff
@@ -13,6 +13,25 @@ import (
...
+
+	str, fwd := onChatMsg(cc, pkt)
+
+	if fwd {
+		cc.server().Send(*&mt.Pkt{
+			Cmd: pkt,
+		})
+	}
+
+	return str, fwd
```
Comment (HimbeerserverDE) [inline chat.go:32]: Why are you returning the processed string? What is it useful for?

**Review by HimbeerserverDE — CHANGES_REQUESTED** [review]
What is all of this needed for anyway? Seems like an irrelevant feature

File: chat.go (hunk 4 above)
Comment (DerZombiiie) [reply chat.go:24]: can't cuz return values

File: chat.go (hunk 6 above)
Comment (DerZombiiie) [reply chat.go:32]: its the returnvalue of the chatcommand not the string
msg != str

File: chat.go (hunk 4 above)
Comment (HimbeerserverDE) [reply chat.go:24]: Can but I guess it’s not that important for now. Resolving.

File: chat.go (hunk 5 above)
Comment (DerZombiiie) [reply chat.go:20]: no, you actually cant:
```
# github.com/HimbeerserverDE/mt-multiserver-proxy
./chat.go:18:10: github.com/anon55555/mt.ToSrvChatMsg composite literal uses unkeyed fields
```

Comment (HimbeerserverDE) [reply chat.go]: Interesting, mt magic

**Review by HimbeerserverDE — COMMENTED** [review]
The major issues of this PR have been resolved, but there's one thing we still have to talk about before merging. The way it is now the CC will get the output given to SendChatMsg by chat command handlers just like before => The player sees it. The same thing applies to return values of command handlers but they're at least made available to the caller.

How should this work? Redirect all output to the calling plugin or send it to the client like normal?
Redirecting would be quite annoying and dirty.
Continuing like normal would require the return value to be removed, anything else makes no sense.

I would ignore the output and just send it to the player but not the calling plugin.

**Comment by DerZombiiie** [comment]
You're right.
but the complete idea of sending chatMsgs as a client is kinda dirty

**Pull request merged**